🎨 Added searchable dropdown with custom option for publication language

betschki opened the pull request:
Replaced text input with SelectWithOther component to prevent invalid locale entries. Users can select from 64 predefined locales or enter custom BCP 47 compliant codes via "Other" option.

This addresses the issue where users are entering invalid values like "English" instead of proper locale codes.

betschki committed:
🎨 Added searchable dropdown with custom option for
publication language

Replaced text input with SelectWithOther component to prevent invalid locale entries. Users can select from 64 predefined locales or enter custom BCP 47 compliant codes via "Other" option. This addresses the issue where users were entering invalid values like "English" instead of proper locale codes.

betschki committed:
🔧 Enhance BCP 47 validation for publication language input

betschki commented:
Fixing linting and tests with CodeRabbit's comment

betschki committed:
🔧 Refactor locale settings components for improved readability and validation messages

betschki commented:
A few screenshots for reference:

This is how it looks like if a "custom locale" is entered. It would also show up like this if a user has set a "custom locale" in Ghost before this potential merge.
![image](https://github.com/user-attachments/assets/be7243df-c3d9-49ff-b31f-97f5475ef748)

This is the dropdown (searchable):
![image](https://github.com/user-attachments/assets/ee48dd9c-6893-42c8-bcc5-eab81e88001a)

Selecting "Other" from the list switches from the dropdown to a (validated) text input field:
![image](https://github.com/user-attachments/assets/6ccf36c9-57c0-45bc-b1ef-d57c5a7c6600)

The input field is validated against BCP 47:
![image](https://github.com/user-attachments/assets/642e1746-7dd7-4c0a-9663-44898944d03c)

An empty string is also not possible:
![image](https://github.com/user-attachments/assets/80f7cf27-96d5-4b6c-ab0c-5fadfe5a4ce5)

betschki committed:
🎨 Enhance custom input handling in SelectWithOther component with improved back-to-list button

betschki committed:
🎉 Implement locale validation utility and corresponding unit tests

betschki committed:
🔧 Fix validation test for invalid locale formats and ensure error messages are displayed correctly

betschki committed:
🔧 Update locale examples in PublicationLanguage component and validation tests for improved clarity

betschki committed:
🔧 Moved supported locales to i18n package with Typescript support

betschki committed:
🔧 Update placeholder text for custom locale input to improve clarity

betschki committed:
🔧 Clarified purpose of constant

betschki committed:
🔧 Change position for back to list button

betschki committed:
🔧 Improve validation error message visibility for custom locale input

kevinansfield commented on apps/admin-x-settings/src/utils/localeValidation.ts:
```diff
@@ -0,0 +1,149 @@
+        return errorMessage;
+    }
+
+    // Grandfathered irregular tags (exact matches, case insensitive)
+    const grandfatheredIrregular = [
+        'en-gb-oed', 'i-ami', 'i-bnn', 'i-default', 'i-enochian', 'i-hak',
+        'i-klingon', 'i-lux', 'i-mingo', 'i-navajo', 'i-pwn', 'i-tao',
+        'i-tay', 'i-tsu', 'sgn-be-fr', 'sgn-be-nl', 'sgn-ch-de'
+    ];
+    
+    // Grandfathered regular tags (exact matches, case insensitive)
+    const grandfatheredRegular = [
+        'art-lojban', 'cel-gaulish', 'no-bok', 'no-nyn', 'zh-guoyu',
+        'zh-hakka', 'zh-min', 'zh-min-nan', 'zh-xiang'
+    ];
```
What is the source of these grandfathered tags? I think it would be useful to add a bit more explanation to the comments to help future code explorers

kevinansfield reviewed (changes requested):
A couple of minor things added in comments above but otherwise this looks+works great! 🙌

betschki committed:
Add documentation for IANA Language Subtag Registry

betschki committed:
Restructure custom value calculation

betschki committed:
Remove duplicate isCurrentlyCustomValue calculation

betschki commented on apps/admin-x-design-system/src/global/form/SelectWithOther.tsx:
```diff
@@ -97,8 +91,13 @@ const SelectWithOther: React.FC<SelectWithOtherProps> = ({
+    const isCurrentlyCustomValue = selectedValue &&
+        !options.some(opt => opt.value === selectedValue) &&
+        selectedValue !== otherOption.value;
+    
```
Fixed! :+1:

kevinansfield commented:
@betschki thanks for making those changes 🙌  Everything looks good to merge now but the branch has conflicts that need resolving

betschki committed:
🎨 Add SelectWithOther component for publication language settings

- Implement SelectWithOther component supporting dropdown selection with custom text input
- Add comprehensive locale validation with BCP 47 support
- Add searchable locale dropdown with 100+ predefined options
- Move supported locales to i18n package with TypeScript support
- Add comprehensive test coverage for component and validation
- Improve user experience with clear validation messages and smooth transitions

Resolves custom locale input requirements and improves publication language configuration.

betschki committed:
Merge branch 'main' into select-with-other-option-for-publication-language
